### PR TITLE
Fixes #4278 "Media details categories editor: Category not actually added"

### LINKED ISSUE #4278: Media details categories editor: Category not actually added
**Summary:** 

when Add category to any (image,video) it show in add in categories(MedailDetail)
But when re again open current image  medailDetail it not show in categories its mean it is not upload to database  

**Steps to reproduce:** 
See in Screenshot 

**Device and Android version:** 
Android Device:Redmi 4
API :28
**Commons app version:** 
Latest master (3.0.0)
**Screen-shots:** 

https://user-images.githubusercontent.com/65972015/109042967-cdb29f80-76f6-11eb-9b8c-135ff2ae4829.mp4

**Would you like to work on the issue?**
YES
Please let us know whether you want to fix the issue by yourself. If not, anyone can get the issue assigned to them.
YES

### PR DESCRIPTION
**Pull request by 4D17Y4**
**Description (required)**

Fixes #4278

What changes did you make and why?

Token responses were fetched on the main thread causing `android.os.NetworkOnMainThreadException`.
Performed the operation off the main thread to Tackle the issue.

**Tests performed (required)**

Tested betaDebug on Noia 6.1+ with API level 29.

### CODE REVIEW

**Comment by 4D17Y4** [comment]
I'm not sure about the class name, would refractor if necessary.
Up for review @nicolas-raoul, @neslihanturan.

File: data-client/src/main/java/org/wikipedia/csrf/CsrfTokenClient.java [hunk 1]
```diff
@@ -210,4 +214,27 @@ public void twoFactorPrompt() {
     private interface RetryCallback {
         void retry();
     }
+
+    class getCSRFTokenResponse implements Callable<Response<MwQueryResponse>> {
+        private Service service;
```
Comment (nicolas-raoul) [inline data-client/src/main/java/org/wikipedia/csrf/CsrfTokenClient.java:219]: Tiny javadoc

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
Category addition works great with this PR (and did not work before), thanks!

File: data-client/src/main/java/org/wikipedia/csrf/CsrfTokenClient.java [hunk 2]
```diff
@@ -210,4 +214,27 @@ public void twoFactorPrompt() {
     private interface RetryCallback {
         void retry();
     }
+
+    class getCSRFTokenResponse implements Callable<Response<MwQueryResponse>> {
```
[the remark this answers is not in the document]
Comment (4D17Y4) [reply data-client/src/main/java/org/wikipedia/csrf/CsrfTokenClient.java:218]: This is analogous to AsyncTask in java but as AsyncTask is deprecated, I've used java executors to fetch data off the main thread. Should I just call it `CSRFTokenResponse` ?

[the remark this answers is not in the document]
Comment (nicolas-raoul) [reply data-client/src/main/java/org/wikipedia/csrf/CsrfTokenClient.java:218]: How about something like `CsrfTokenCallExecutor`?
By the way CSRF is best written Csrf as the wikipedia class uses that case (there is no strict convention but 4 letters acronyms are [usually](https://stackoverflow.com/questions/2236807/java-naming-convention-with-acronyms) written this way).

[the remark this answers is not in the document]
Comment (4D17Y4) [reply data-client/src/main/java/org/wikipedia/csrf/CsrfTokenClient.java]: Done.
I'll keep that in mind.
Thanks.

File: data-client/src/main/java/org/wikipedia/csrf/CsrfTokenClient.java [hunk 3]
```diff
@@ -210,4 +214,30 @@ public void twoFactorPrompt() {
     private interface RetryCallback {
         void retry();
     }
+
+    /**
+     * Class CsrfTokenCallExecutor which implement callable interface to get CsrfTokenCall.
+     */
+    class CsrfTokenCallExecutor implements Callable<Response<MwQueryResponse>> {
+        private Service service;
```
Comment (nicolas-raoul) [inline data-client/src/main/java/org/wikipedia/csrf/CsrfTokenClient.java:222]: javadoc for this member, for instance say in 5 to 10 words what this service offers.

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]

File: data-client/src/main/java/org/wikipedia/csrf/CsrfTokenClient.java (hunk 3 above)
Comment (4D17Y4) [reply data-client/src/main/java/org/wikipedia/csrf/CsrfTokenClient.java:226]: Done.

**Review by neslihanturan — APPROVED** [review]
Thanks @4D17Y4 , I merge this one since all of the reviews of @nicolas-raoul seems done and PR works like a charm.

**Pull request merged**